src/Gadgets.ts

```diff
@@ -0,0 +1,110 @@
+import Downloader from './Downloader.js'
+import * as logger from './Logger.js'
+import MediaWiki from './MediaWiki.js'
+
+class Gadgets {
+  private static instance: Gadgets
+  private gadgets: Gadget[] | undefined
+
+  public static getInstance() {
+    if (!Gadgets.instance) {
+      Gadgets.instance = new Gadgets()
+    }
+    return Gadgets.instance
+  }
+
+  public async fetchGadgets() {
+    const gadgetsUrl = Downloader.apiUrlDirector.buildGadgetsUrl()
+    const gadgetsResponse = await Downloader.getJSON<GadgetQueryResult>(gadgetsUrl)
+    if (!gadgetsResponse.batchcomplete) {
+      throw new Error(`Error while fetching gadgets with ${gadgetsUrl}, batch seems to be incomplete, scraper does not expect/support pagination here`)
+    }
+    this.gadgets = gadgetsResponse.query.gadgets
```

Comment:
As the query only received warnings with no data returned, the `query` key is already missing (see the JSON above). So your current check if gadgets is installed would still throw a `TypeError: Cannot read properties of undefined (reading 'gadgets')`

```suggestion
    this.gadgets = gadgetsResponse.query && gadgetsResponse.query.gadgets
```

(After #2385 we would always have the `query` key (because siteinfo always exists) and only the `gadgets` key missing.)